File: HOT100/13.堆/347.前-k-个高频元素.cpp
/*
 * @lc app=leetcode.cn id=347 lang=cpp
 *
 * [347] 前 K 个高频元素
 */
#include "tools.h"
// @lc code=start
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> hash;
        int max_cnt = 0;
        for (auto x : nums)
        {
            hash[x] ++;
            max_cnt = max(max_cnt, hash[x]);
        }
        vector<vector<int>> bucket(max_cnt + 1);
        for (auto& [key, val] : hash)
        {
            bucket[val].push_back(key);
        }
        vector<int> res;
        for (int i = max_cnt; i >= 0; i -- )
        {
            res.insert(res.end(), bucket[i].begin(), bucket[i].end());
            k -= bucket[i].size();
            if (k <= 0) break;
        }
        return res;

    }
};
// @lc code=end

File: HOT100/13.堆/347.前-k-个高频元素.py
#
# @lc app=leetcode.cn id=347 lang=python3
#
# [347] 前 K 个高频元素
#
from mytools import *
# @lc code=start
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        hash = defaultdict(int)
        for x in nums:
            hash[x] += 1
        h = []
        for key, val in hash.items():
            heapq.heappush(h, (val, key))
            if len(h) > k: heapq.heappop(h)

        return [x for _, x in h]

# @lc code=end

